sklearn/tree/_tree.pyx

```diff
@@ -109,16 +108,14 @@ cdef class TreeBuilder:
             # since we have to copy we will make it fortran for efficiency
             X = np.asfortranarray(X, dtype=DTYPE)
 
-        if y.dtype != DOUBLE or not y.flags.contiguous:
-            y = np.ascontiguousarray(y, dtype=DOUBLE)
-
         if (sample_weight is not None and
-            (sample_weight.dtype != DOUBLE or
-            not sample_weight.flags.contiguous)):
+            (sample_weight.base.dtype != DOUBLE or
+            not sample_weight.base.flags.contiguous)):
                 sample_weight = np.asarray(sample_weight, dtype=DOUBLE,
                                            order="C")
 
-        return X, y, sample_weight
+        sample_weight_exists = sample_weight is not None and sample_weight.size > 0
+        return X, sample_weight.base if sample_weight_exists else None
```

Comment:
I don't understand the point of these changes. Could you explain what motivates them?

Reply:
Sample weight can be None or empty. In that case the base attribute is not defined.

Reply:
What I meant is that I don't think we need to convert it back to an `np.ndarray` at all.

In the scenario where we need to do so, returning `np.asarray(sample_weight)` is preferable IMO.

Reply:
Oh right. Yes I think we can directly return the memory views.